### PR TITLE
Add support for automatic database users for Postgres

### PR DESCRIPTION
**Pull request by r0mant**
Implements [RFD 113](https://github.com/gravitational/teleport/blob/master/rfd/0113-automatic-database-users.md).

A couple of minor optional things mentioned in the RFD are missing from the implementation (like auditing activate/deactivate queries). I'll add them later while doing more testing during review or as a follow up.

### CODE REVIEW

File: lib/services/role.go [hunk 1]
```diff
@@ -1414,6 +1419,56 @@ func (set RoleSet) AdjustDisconnectExpiredCert(disconnect bool) bool {
 	return disconnect
 }
 
+// CheckDatabaseRoles returns whether a user should be auto-created in the
+// database and a list of database roles to assign.
+func (set RoleSet) CheckDatabaseRoles(database types.Database) (create bool, roles []string, err error) {
```
Comment (greedy52) [inline lib/services/role.go:1424]: should `database.GetAdminUser()` be checked here or somewhere? assuming a database that matches the role labels but has NO Admin user shouldn't create users.

File: lib/srv/db/postgres/users.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	conn, err := e.pgxConnect(ctx, sessionCtx.WithUser(sessionCtx.Database.GetAdminUser()))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer conn.Close(ctx)
+
+	// We could call this once when the database is being initialized but
+	// doing it here has a nice "self-healing" property in case the Teleport
+	// bookkeeping group or stored procedures get deleted or changed offband.
+	err = e.initAutoUsers(ctx, conn)
```
Comment (greedy52) [inline lib/srv/db/postgres/users.go:46]: Database agent is usually close to the database server but three calls may add up. Do we have a rough estimate of how much latency this adds?

**Review by greedy52 — COMMENTED** [review]
First round. Will test it out next round.

**Review by smallinsky — COMMENTED** [review]
First pass.

File: lib/srv/db/common/role/role.go [hunk 3]
```diff
@@ -23,14 +23,19 @@ import (
 )
 
 // DatabaseRoleMatchers returns role matchers based on the database.
-func DatabaseRoleMatchers(db types.Database, user, database string) services.RoleMatchers {
-	roleMatchers := services.RoleMatchers{
-		services.NewDatabaseUserMatcher(db, user),
+func DatabaseRoleMatchers(db types.Database, user, database string, roles ...string) services.RoleMatchers {
```
[the remark this answers is not in the document]
Comment (r0mant) [reply lib/srv/db/common/role/role.go:26]: Added the logic for passing the roles there as well.

File: lib/srv/db/postgres/users.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer conn.Close(ctx)
+
+	e.Log.Infof("Deactivating PostgreSQL user %q.", sessionCtx.DatabaseUser)
+
+	_, err = conn.Exec(ctx, deactivateQuery, sessionCtx.DatabaseUser)
+	if err != nil {
+		return trace.Wrap(err)
```
[the remark this answers is not in the document]
Comment (r0mant) [reply lib/srv/db/postgres/users.go:86]: It does not.

File: lib/services/role.go (hunk 1 above)
Comment (r0mant) [reply lib/services/role.go:1425]: I'm not sure it belongs here tbh. It is checked as part of the user provisioning engine:

https://github.com/gravitational/teleport/blob/a1630b11f53c6460469e0669950ff1068c214918/lib/srv/db/postgres/users.go#L33-L35

File: lib/srv/db/postgres/users.go (hunk 2 above)
Comment (r0mant) [reply lib/srv/db/postgres/users.go:46]: Yeah, well, we're already adding basically another connection roundtrip for each auto-provisioned user, right? I don't think added latency for 3 extra queries here is huge but we can see if it becomes a problem and optimize then?

File: lib/srv/db/common/interfaces.go [hunk 5]
```diff
@@ -84,3 +84,12 @@ type Engine interface {
 type Users interface {
 	GetPassword(ctx context.Context, database types.Database, userName string) (string, error)
 }
+
+// AutoUsersBackend defines an interface for automatic user provisioning
+// a particular database engine should implement.
+type AutoUsersBackend interface {
```
[the remark this answers is not in the document]
Comment (r0mant) [reply lib/srv/db/common/interfaces.go]: Renamed this interface to just `AuthUsers` and the struct with the same name to `UserProvisioner`.

File: lib/srv/db/postgres/engine.go [hunk 6]
```diff
@@ -197,21 +211,28 @@ func (e *Engine) handleStartup(client *pgproto3.Backend, sessionCtx *common.Sess
...
 
 func (e *Engine) checkAccess(ctx context.Context, sessionCtx *common.Session) error {
+	// When using auto-provisioning, force the database username to be same
+	// as Teleport username. If it's not provided explicitly, some database
+	// clients (e.g. psql) get confused and display incorrect username.
+	if sessionCtx.AutoCreateUser {
+		if sessionCtx.DatabaseUser != sessionCtx.Identity.Username {
+			return trace.AccessDenied("please use your Teleport username (%q) to connect instead of %q",
+				sessionCtx.Identity.Username, sessionCtx.DatabaseUser)
+		}
```
[the remark this answers is not in the document]
Comment (r0mant) [reply lib/srv/db/postgres/engine.go:221]: Yeah, I had a similar idea. I've updated `tsh db connect` to default to Teleport username if we detect the database has admin user defined which likely means auto-users are being used.

**Comment by r0mant** [comment]
@smallinsky @greedy52 I've addressed your feedback folks, mind taking another look?

File: lib/services/role.go (hunk 1 above)
Comment (greedy52) [reply lib/services/role.go:1425]: For example, say multiple Postgres databases match the same role and some have admin user set but others don't. Then the ones without admin user will always fail. Unless this is intended behaviour where all databases matching this role must have admin user set.

File: lib/auth/auth.go [hunk 7]
```diff
@@ -4756,10 +4756,20 @@ func (a *Server) isMFARequired(ctx context.Context, checker services.AccessCheck
 			return nil, trace.Wrap(notFoundErr)
 		}
 
+		autoCreate, databaseRoles, err := checker.CheckDatabaseRoles(db)
```
Comment (greedy52) [inline lib/auth/auth.go:4759]: similar idea to the other comment, for a database that doesn't have admin user set, or a database that is not even postgres, as long this database matches this auto-create role, database-user matcher will be skipped. It's ok if this is designed/intended behaviour. Though users have to be very careful when crafting these auth-create roles.

**Review by greedy52 — APPROVED** [review]
tested with RDS today and this feature is awesome!

One small nit on a corner case, when admin user is not setup properly, message from tsh is not very obvious what's wrong sometimes:
```
$ tsh db connect --db-user STeve --db-name test postgres-rds
psql: error: connection to server at "localhost" (::1), port 58348 failed: Connection refused
	Is the server running on that host and accepting TCP/IP connections?
connection to server at "localhost" (127.0.0.1), port 58348 failed: FATAL:  password authentication failed for user "not-found"
ERROR: exit status 2
```

File: lib/services/role.go (hunk 1 above)
Comment (r0mant) [reply lib/services/role.go:1425]: Yeah, I think this would be expected behavior, right? Your role requires automatic user provisioning so if the database doesn't have an admin user configured it should fail.

File: lib/srv/db/postgres/activate-user.sql [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+        execute format('alter user %I with login', username);
+    else
+        execute format('create user %I in role "teleport-auto-user"', username);
+    end if;
+    -- Assign all roles to the created/activated user.
+    foreach role_ in array roles
+    loop
+        execute format('grant %I to %I', role_, username);
+    end loop;
+end;$$;
```
Comment (smallinsky) [inline lib/srv/db/postgres/activate-user.sql:28]: I'm wonder about case where the same user already exists in database and overlaps with teleport username.  
I that case it can happen that current internal user permission that overlaps with teleport username will be escalated to the teleport access.

File: lib/services/role_test.go [hunk 9]
```diff
@@ -3668,6 +3678,112 @@ func mustMakeTestWindowsDesktop(labels map[string]string) types.WindowsDesktop {
 	return d
 }
 
+func TestCheckDatabaseRoles(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (r0mant) [reply lib/services/role_test.go:3691]: Yes, good call. Updated.

File: lib/srv/db/postgres/activate-user.sql (hunk 8 above)
Comment (r0mant) [reply lib/srv/db/postgres/activate-user.sql:28]: Existing user won't receive extra permissions. The "create user" call will fail with "already exists" error so the procedure will end early and "grant" statements won't be executed.

I've also updated this scenario to return an error instead of letting user to log in, which is how I originally thought about handling it in the RFD. So the behavior is:

```
➜  ~ tsh db connect --db-name=postgres local
psql: error: connection to server at "root.gravitational.io" (127.0.0.1), port 3080 failed:
				your Teleport user requires automatic database user provisioning
				but user "roman@goteleport.com" already exists in this PostgreSQL database and is
				not managed by Teleport
ERROR: exit status 2
```

**Comment by r0mant** [comment]
@smallinsky PTAL?

**Review by smallinsky — APPROVED** [review]

**Comment by webvictim** [comment]
@r0mant Please see related issue about support not also being added to the Terraform provider: https://github.com/gravitational/teleport-plugins/issues/870
